src/index.js

```diff
@@ -1,390 +1,586 @@
-      ) {
-        updateAttribute(
-          element,
-          name,
-          attributes[name],
-          oldAttributes[name],
-          isSvg
-        )
-      }
-    }
+var recycleElement = function(element) {
+  return createVNode(
+    element.nodeName.toLowerCase(),
+    EMPTY_OBJECT,
+    map.call(element.childNodes, recycleChild),
+    element,
+    null,
+    RECYCLED_NODE
+  )
+}
 
-    var cb = isRecycling ? attributes.oncreate : attributes.onupdate
-    if (cb) {
-      lifecycle.push(function() {
-        cb(element, oldAttributes)
-      })
-    }
-  }
+var patch = function(container, element, lastNode, nextNode, eventProxy) {
+  var lifecycle = []
 
-  function removeChildren(element, node) {
-    var attributes = node.attributes
-    if (attributes) {
-      for (var i = 0; i < node.children.length; i++) {
-        removeChildren(element.childNodes[i], node.children[i])
-      }
+  patchElement(container, element, lastNode, nextNode, lifecycle, eventProxy)
 
-      if (attributes.ondestroy) {
-        attributes.ondestroy(element)
-      }
-    }
-    return element
-  }
+  while (lifecycle.length > 0) lifecycle.pop()()
 
-  function removeElement(parent, element, node) {
-    function done() {
-      parent.removeChild(removeChildren(element, node))
-    }
+  return nextNode
+}
 
-    var cb = node.attributes && node.attributes.onremove
-    if (cb) {
-      cb(element, done)
-    } else {
-      done()
+export var h = function(name, props) {
+  var node
+  var rest = []
+  var children = []
+  var length = arguments.length
+
+  while (length-- > 2) rest.push(arguments[length])
+
+  if ((props = props == null ? {} : props).children != null) {
+    if (rest.length <= 0) {
+      rest.push(props.children)
     }
+    delete props.children
   }
 
-  function patch(parent, element, oldNode, node, isSvg) {
-    if (node === oldNode) {
-    } else if (oldNode == null || oldNode.nodeName !== node.nodeName) {
-      var newElement = createElement(node, isSvg)
-      parent.insertBefore(newElement, element)
-
-      if (oldNode != null) {
-        removeElement(parent, element, oldNode)
+  while (rest.length > 0) {
+    if (isArray((node = rest.pop()))) {
+      for (length = node.length; length-- > 0; ) {
+        rest.push(node[length])
       }
-
-      element = newElement
-    } else if (oldNode.nodeName == null) {
-      element.nodeValue = node
+    } else if (node === false || node === true || node == null) {
     } else {
-      updateElement(
-        element,
-        oldNode.attributes,
-        node.attributes,
-        (isSvg = isSvg || node.nodeName === "svg")
-      )
+      children.push(typeof node === "object" ? node : createTextVNode(node))
+    }
+  }
 
-      var oldKeyed = {}
-      var newKeyed = {}
-      var oldElements = []
-      var oldChildren = oldNode.children
-      var children = node.children
+  return typeof name === "function"
+    ? name(props, (props.children = children))
```

Comment:
A Component will received children property and children argument ?

Reply:
Good catch. It's an implementation detail. If you need `children` for some reason, use `props.children`. I am still open to change this.

Reply:
@frenzzy

Reply:
`props.children` definitely cooler then the second argument, see pros and cons here: https://github.com/hyperapp/hyperapp/issues/569
I am worried only about performance and memoization:
https://github.com/hyperapp/hyperapp/blob/2b3d85625321038ed2369ddb4ecc320a9e20984e/src/index.js#L458

If everything is ok with `props.children` would be nice to remove `children` from the second argument IMO.

Reply:
```js
const Component = (props) => <div>{props.children}</div>

app({
  init: { children: 'Hello' },
  view: (state) => (
    <main>
      <Component {...state}>World</Component>
      <div>{state.children}</div>
    </main>
  ),
  container: document.body,
})
```
Can you guess what this will render without opening the demo?
1. `<main><div>Hello</div><div>World</div></main>`
2. `<main><div>Hello</div><div>Hello</div></main>`
3. `<main><div>World</div><div>World</div></main>`
4. throws an error

Demo: https://codepen.io/frenzzy/pen/Zjbqpq/left/?editors=0010

Reply:
@frenzzy So, don't use `children` in your state?

Reply:
lol, v2 allows everything in state but you should not use `children` key in it to prevent internal state mutation by hyperapp components implementation...

It is pretty common when an api returns data with `children` keys, for example for tree structure of pages, nav items, breadcrumbs, etc. It is not cool when you need to change api as a workaround for hyperapp.

If you still want to disallow `children` key, maybe it must throw an error instead of silently mutating state? Because it will inevitably lead to difficult to perceptible or hidden/uncatched errors in UI.

Reply:
@frenzzy So you'd prefer to drop ({children,...props}) and go back to (props, children)?

Reply:
yes, because it is less hacky implementation, we already discussed this in #569

Reply:
👌

Reply:
Any alternatives? Why can preact get away with the same hack? 🤔

Reply:
Purity forever! Say NO to mutations :D

Reply:
Okay, so to clarify, we're back to (props, children)?

Reply:
Yes!
- no mutations
- no pitfalls for `children` key
- simple memoization with `===`